### DISCUSSION TITLE
Redirect and Set-Cookie in next 13 (new route.js)

### DISCUSSION THREAD

**Discussion by imfelipedev** [vercel/next.js#48434] [Help]
### Summary

I'm developing an auth api, using the new "app" folder route system. But I realized that it is not possible to redirect and set a cookie and vice versa. (my folder: "app/api/auth/route.js").

This is my current return, I want to set the cookie and redirect the client to "/dashboard"
```
const sessionData = await session.create(accountCache.id);
return NextResponse.json(
	{
		success: true
	},
	{
		status: 201,
		headers: { 'Set-Cookie': `session_id=${sessionData.token}; HttpOnly; path=/; Max-Age=2592000; Secure` }
	}
);
```

**Comment by TravisBallard** [comment]
Were you able to resolve this at all? I'm running into the same issue in a project that i am working on

**Reply by imfelipedev** [reply]
Hello! No, when I was developing this solution, I looked a lot in the documentation and didn't find anything about it.

**Comment by PaulMest** [comment]
Here is a snippet that works for my use case. Have you tried something like this?

```TypeScript
// route.ts
export async function POST(request: Request) {
  const formData = await request.formData();
  const password = formData.get("password");

  const destinationUrl = new URL("/dashboard", new URL(request.url).origin);
  const response = NextResponse.redirect(destinationUrl, { status: 302 });

  if (checkPassword(password)) {
    response.cookies.set("login", process.env.LOGIN_CODE ?? "true", {
      path: "/",
      httpOnly: true,
      maxAge: 60 * 60 * 24 * 7, // 1 week
    });
  }

  return response;
}
```

**Reply by imfelipedev** [reply]
I tried to use your method but it did not redirect

**Reply by PaulMest** [reply]
@zFelpszada Did you include `{status: 302}`? https://github.com/vercel/next.js/discussions/48434#discussioncomment-6336966

**Comment by lukemorton** [comment]
I am currently having the same issue. The redirect appears to happen in the browser network console but then the destination page spins/loads indefinitely.

``` ts
export async function POST(request: Request) {
  const data = await request.formData();
  const credential = data.get("credential") as string;
  const { sessionId } = await SignInWithGoogle({ credential });

  const response = NextResponse.redirect(
    "http://localhost:3000/properties/add"
  );

  response.cookies.set("sessionId", sessionId, {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
  });

  return response;
}
```

**Reply by PaulMest** [reply]
In my experience, without explicitly adding `{ status: 302 }`, the request would just hang and the browser wouldn't redirect. So try adding that as the second parameter here...

```typescript
  const response = NextResponse.redirect(
    "http://localhost:3000/properties/add",
    { status: 302 }
  );
```

**Reply by lukemorton** [reply]
That's resolved it, thanks Paul! I owe you :)

**Comment by tonywei92** [comment]
I added `{ status: 302 }` still just hang

**Comment by gurvirbaraich** [comment]
I hope this works for you, if you have a better solution let me know!.

```js
// Sending a html script as response from the server
return new Response("<script>window.location.href = '/'</script>", {
   status: 200,
   headers: {
      "Content-Type": "text/html", // Setting the response type as text/html
      "Set-Cookie": serialize(COOKIE_NAME, token, {
         path: "/",
         httpOnly: true,
         sameSite: "strict",
         maxAge: COOKIE_MAX_AGE,
         secure: process.env.NODE_ENV === "production",
      }),
   },
});
```

**Reply by bharat5604** [reply]
i am having issue with updating cookie value from route handler

**Reply by wisnuprama** [reply]
I found another problem and this is a working workaround solution. Thanks!

My case only happen when

1. We have a link to set cookie based on params e.g. https://example.com?this_is_cookie=true
2. Then in route.js, we set cookie
    ```
    function GET() {
         const r = NextResponse.redirect("/")
         r.cookies.setCookie('this_is_cookie', searchParams.this_is_cookie)
         return r
    }
    ```

The problem only come when users 

1. Copy paste that link in an Android app
2. Click the link and the app trigger Chrome browser (not webview)
3. The redirection request does not have cookies, based on manual logs in App router page
4. Return the page not based on cookies (unexpected). However, upon inspect element, I can see the cookies
5. Users trigger pull to refresh
6. It will have cookies and display correctly based on the cookies

So somehow there is a gap between step 2 - 5. My feeling the redirection request actually have cookies because in step 5-6, users can see the actual content based on cookies eventually. So this workaround solution provided by @gurvirbaraich works because it is mechanical trigger of step 5-6.

**Reply by TheRafaelFarias** [reply]
This worked perfectly for me, it redirected and passed cookies as it should.

For some reason using `cookies()` or `response.cookies.set()` doesn't work, it passed the `Set-Cookie` but the redirected path didn't contain any cookie

**Comment by Aelghani** [comment]
Here is a snippet that works for my use case. if you use app router it will sure work for you.

```
import { NextRequest, NextResponse } from "next/server";

export async function POST(request: NextRequest) {
  try {
    const reqBody = await request.json();
    
    const tokenValue = {
        id: 'userID',
        username: 'username'
    }

    const response = NextResponse.json({
      message: "Login successful",
      success: true,
    });

    response.cookies.set("token", tokenValue, {
      httpOnly: true,
    });

    return response;
  } catch (error: any) {
    return NextResponse.json({ error: error.message }, { status: 500 });
  }
}

```

**Comment by luskin** [comment]
We are struggling with this as well. The following code properly sets the cookie, but the render of the redirected-to page doesn't have access to it until the page is refreshed. Any advice?

```ts
const nextResponse = NextResponse.redirect(redirectUrl)

if (response.session) {
  nextResponse.cookies.set("__session", response.session.token, {
    expires: new Date(response.session.expiresAt),
    httpOnly: true,
    secure: true,
    path: "/",
    sameSite: "strict",
  })
}
return nextResponse
```

**Reply by luskin** [reply]
🚨 Setting `sameSite: "lax"` resolved this issue for us. There are other security concerns to consider with that, but this ultimately passes the cookie properly to the redirected page for immediate use in render.

**Reply by markedwards** [reply]
Had the same issue.  In our case, the route handler is doing intermediate processing of authentication for a password reset.  Specifically, a Stytch URL is redirecting to this route, which needs to set the cookie before redirecting again to the password reset page.

Is the need to set "lax" due to the external domain in the chain?  Or is this an issue in Next?

**Reply by jericopulvera** [reply]
works wonders. this needs to be documented somehow.

**Reply by markedwards** [reply]
Pretty sure this is just a limitation of cookies, nothing to do with Next.

**Reply by jericopulvera** [reply]
I agree. But might be a good idea to note this at cookie or redirect api documentation of Next since it's a common use case.

**Reply by Arhan13** [reply]
Yeah this should surely be mentioned on the docs, I have wasted so many hours because of this issue. Thanks @luskin for pointing it out

**Reply by qhkm** [reply]
Faced this same issue and set the sameSite value to lax fix the issue. Agree this needs to be documented

**Comment by AssafMeidan** [comment]
Adding a related finding for anyone dealing with Set-Cookie + redirect issues in App Router:

**React hydration intercepts `<form method="POST">` submissions from Next.js pages** — even Server Components get hydrated, converting native form POSTs into `fetch()` calls. Browsers do **not** store `Set-Cookie` headers from `fetch()` responses (per the Fetch spec).

This means if your login form lives in a Next.js page component, the browser will never store the session cookie — even though the server sends `Set-Cookie` correctly.

**Workaround:** Serve the login form as raw HTML from a route handler (`GET /api/auth/login`). No React, no JS = native form POST = browser stores the cookie normally. The POST handler returns `200` with `Set-Cookie` + `<meta http-equiv="refresh" content="0;url=/dashboard">` for the redirect (instead of 303, which can race with cookie storage).

This is a separate issue from the 303-redirect-drops-Set-Cookie problem, but it compounds it — you can fix the redirect behavior and still lose cookies because of the hydration interception.

Full write-up: https://github.com/vercel/next.js/discussions/60581#discussioncomment-15840971
